Tests/SwiftLintFrameworkTests/TestHelpers.swift

```diff
@@ -58,10 +58,9 @@ private func render(locations: [Location], in contents: String) -> String {
     var contents = contents.bridge().lines().map { $0.content }
     for location in locations.sorted(by: > ) {
         guard let line = location.line, let character = location.character else { continue }
-        var content = contents[line - 1]
-        let index = content.index(content.startIndex, offsetBy: character - 1)
-        content.insert("↓", at: index)
-        contents[line - 1] = content
+        let content = NSMutableString(string: contents[line - 1])
```

Comment:
@marcelofabri, `Location.character` has always been an `Int` that describes the distance into a string in **`UTF‐16` units**. It was probably named `character` based off of `NSString`.

Before my change, this method measured off that distance into the string in **extended grapheme clusters** instead, which Swift calls `Character`, and tried to insert `↓` at that location to mark the location of the failure.

While both measurements have been given the name “character”, they are not the same thing.

For example, “ơ̧̈̂” is only `1` grapheme cluster but `5` UTF‐16 units (o +  ̛ +  ̧ +  ̈ +  ̂).
Also, “👨‍👩‍👧‍👦123” is only `4` grapheme clusters but `5` UTF‐16 units (a pair of surrogates, 1, 2, 3).

(The following example offsets are zero‐based.)

Using mis‐matched offsets can always result in an unexpected position, but whenever an offset points to a UTF‐16 index near the end, there is also the risk of an index out of bounds error. For example, `4` is the UTF‐16 offset of the  ̂ above, but the same offset of `4`, measured in grapheme clusters, is way out of bounds, so the tests crash.
Using the emoji example above, the UTF‐16 offset of `3` points at “2”, but requesting the `3`rd grapheme cluster results in “3” (wrong but no crash). The UTF‐16 offset of `4` points at “3”, but requesting the `4`th cluster grasps at out of bounds (crash).

My fix simply uses the `NSString`‐based, UTF‐16, `Int` offset on—`NSString`. No more mis‐match. I could have used `String.UTF16View`, but that would have incurred unnecessary parsing up and down the string to convert the `Int` to a `String.UTF16Index`.

When @sammy-SC first asked about the out of bounds errors, I assumed he just needed an explanation of Unicode. Later I looked closer and realized the errors had nothing to do with him. It was merely that with the comma nearly always being near the end of the string, this rule is particularly prone to this type of bug, and Samuel just happened the first person the bug messed with. At that point I felt bad and stepped in to untangle it for him. Aside from these string indexing bug fixes, the pull request is all his.

Is that explanation enough?